src/languages/deno/deno.rs

```diff
@@ -53,44 +27,74 @@ impl LanguageImpl for Deno {
+        let deno_dir = store.tools_path(ToolBucket::Deno);
+        let installer = DenoInstaller::new(deno_dir);
+
+        let (deno_request, allows_download) = match &hook.language_request {
+            LanguageRequest::Any { system_only } => (&DenoRequest::Any, !system_only),
+            LanguageRequest::Deno(deno_request) => (deno_request, true),
+            _ => unreachable!(),
+        };
+
+        let deno = installer
+            .install(store, deno_request, allows_download)
+            .await
+            .context("Failed to install deno")?;
+
         // Create env for isolated dependencies
         let mut info = InstallInfo::new(
             hook.language,
             hook.dependencies().clone(),
             &store.hooks_dir(),
         )?;
 
-        let deno_dir = store.cache_path(CacheBucket::Deno);
-        fs_err::tokio::create_dir_all(&deno_dir).await?;
-
-        let DenoInfo {
-            version: deno_version,
-            executable: deno_executable,
-        } = query_deno_info().await?;
-
-        // TODO: how to install the local repo?
-        // if let Some(repo_path) = hook.repo_path() {
-        //     Cmd::new(&deno_executable, "deno add")
-        //         .current_dir(&info.env_path)
-        //         .env(EnvVars::DENO_DIR, &deno_dir)
-        //         .arg("add")
-        //         .arg(format!("npm:{}", crate::fs::relative_to(repo_path, &info.env_path)?.display()))
-        //         .check(true)
-        //         .output()
-        //         .await?;
-        // }
-
-        Cmd::new(&deno_executable, "deno add")
-            .current_dir(&info.env_path)
-            .env(EnvVars::DENO_DIR, &deno_dir)
-            .arg("add")
-            .args(&hook.additional_dependencies)
-            .check(true)
-            .output()
-            .await?;
-
-        info.with_language_version(deno_version)
-            .with_toolchain(deno_executable);
+        info.with_toolchain(deno.deno().to_path_buf());
+        info.with_language_version((**deno.version()).clone());
+
+        let deno_cache_dir = store.cache_path(CacheBucket::Deno);
+        fs_err::tokio::create_dir_all(&deno_cache_dir).await?;
+
+        // Initialize deno.json if we have dependencies to install
+        if hook.repo_path().is_some() || !hook.additional_dependencies.is_empty() {
+            let deno_json = info.env_path.join("deno.json");
+
+            // Check if repo has deno.json to copy, otherwise create minimal one
+            let mut needs_deno_json = true;
+            if let Some(repo_path) = hook.repo_path() {
+                let repo_deno_json = repo_path.join("deno.json");
+                if repo_deno_json.exists() {
+                    // Copy the deno.json from the repo
+                    fs_err::tokio::copy(repo_deno_json, &deno_json).await?;
+                    needs_deno_json = false;
+                }
+                // Deno can run scripts directly from the repo without installation
```

Comment:
I think we need to cache dependencies from `deno.json` (by `deno install`?). The idea is that `install` should set up everything needed for the hooks to run, so when we call `run`, it can work completely offline.

Reply:
not an expert of `deno`, just a user. but i think `deno` maintains it's own global cache of every package that was ever imported in an executed script (be it from NPM, JSR or directly via URL from GitHub)